### PR TITLE
Default to the medium code model in x86 linux

### PR DESCRIPTION
**Pull request by gbaraldi**
This shouldn't have any cost on smaller images because the only thing that gets put into ldata is the system image data, which is only reference via `dlsym`. This allows for images larger than 2gb (tested by putting a 2gb array in the base image)

I did not test how this might be handled in other platforms (windows doesn't support it)

We may want to backport to 1.10/1.11

### CODE REVIEW

File: src/aotcompile.cpp [hunk 1]
```diff
@@ -1655,6 +1657,8 @@ void jl_dump_native_impl(void *native_code,
                                      GlobalVariable::ExternalLinkage,
                                      data, "jl_system_image_data");
         sysdata->setAlignment(Align(64));
+        if (TheTriple.isX86() && TheTriple.isArch64Bit() && TheTriple.isOSLinux())
+            sysdata->setSection(".ldata");
```
Comment (vtjnash) [inline src/aotcompile.cpp:1661]: IIUC, this is already automatic, once you select Medium code model globally, although can also be opted for explicitly for each variable:
https://llvm.org/doxygen/TargetMachine_8cpp_source.html
```suggestion
        sysdata->setCodeModel(CodeModel::Large);
```

Comment (gbaraldi) [reply src/aotcompile.cpp:1661]: It wasn't doing it automatically for me :(. Not sure why

Comment (gbaraldi) [reply src/aotcompile.cpp:1661]: https://github.com/llvm/llvm-project/pull/72078 also the per global option was just added.

Comment (vtjnash) [reply src/aotcompile.cpp:1661]: I guess this should have an LLVM_VERSION ifdef here then?

**Comment by ven-k** [comment]
Can this be backported to 1.10?

File: src/aotcompile.cpp [hunk 2]
```diff
@@ -1655,6 +1657,12 @@ void jl_dump_native_impl(void *native_code,
                                      GlobalVariable::ExternalLinkage,
                                      data, "jl_system_image_data");
         sysdata->setAlignment(Align(64));
+        if (TheTriple.isX86() && TheTriple.isArch64Bit() && TheTriple.isOSLinux())
+#if JL_LLVM_VERSION >= 180000
+            sysdata->setCodeModel(CodeModel::Large);
+#else
```
Comment (vtjnash) [inline src/aotcompile.cpp:1663]:
```suggestion
#if JL_LLVM_VERSION >= 180000
        sysdata->setCodeModel(CodeModel::Large);
#else
        if (TheTriple.isX86() && TheTriple.isArch64Bit() && TheTriple.isOSLinux())
```

**Comment by maleadt** [comment]
This PR breaks precompilation of a bunch of packages, such as LLVM:

```
Precompiling project...
lld: error: relocation R_X86_64_64 cannot be used against symbol 'jlcapi_handle_error_4257'; recompile with -fPIC
>>> defined in /home/tim/.julia/compiled/v1.10/LLVM/jl_bPThkr(text#1.o)
>>> referenced by libLLVM.jl:232 (/home/tim/.julia/packages/LLVM/bzSzE/lib/15/libLLVM.jl:232)
>>>               text#4.o:(julia___init___4235) in archive /home/tim/.julia/compiled/v1.10/LLVM/jl_bPThkr

lld: error: relocation R_X86_64_64 cannot be used against symbol 'jlcapi_handle_diagnostic_4271'; recompile with -fPIC
>>> defined in /home/tim/.julia/compiled/v1.10/LLVM/jl_bPThkr(text#5.o)
>>> referenced by libLLVM.jl:580 (/home/tim/.julia/packages/LLVM/bzSzE/lib/15/libLLVM.jl:580)
>>>               text#4.o:(julia___init___4235) in archive /home/tim/.julia/compiled/v1.10/LLVM/jl_bPThkr
```

The functions mentioned here all define `@cfunction` callbacks. This only seems to happen on the 1.10 backports branch, so probably related to LLVM 15 and fixed in 16?

**Pull request merged**